btb/hyper_parameter.py

```diff
@@ -21,37 +21,64 @@ class ParamTypes(object):
 
 # HyperParameter object
 class HyperParameter(object):
-    def __new__(cls, typ, rang):
-        if cls is HyperParameter:
-            return super(HyperParameter, cls).__new__(CLASS_GENERATOR[typ])
-        else:
-            return super(HyperParameter, cls).__new__(cls)
-
-    def __init__(self, rang, cast):
-        for i, val in enumerate(rang):
-            if val is None:
-                # the value None is allowed for every parameter type
-                continue
-            rang[i] = cast(val)
+
+    param_type = None
+    is_integer = False
+
+    _subclasses = []
+
+    @classmethod
+    def _get_subclasses(cls):
+        subclasses = []
+        for subclass in cls.__subclasses__():
+            subclasses.append(subclass)
+            subclasses.extend(subclass._get_subclasses())
+
+        return subclasses
+
+    @classmethod
+    def subclasses(cls):
+        if not cls._subclasses:
+            cls._subclasses = cls._get_subclasses()
+
+        return cls._subclasses
+
+    def __new__(cls, param_type=None, rang=None):
+        if param_type:
```

Comment:
I personally prefer not to check against an exact value (in this case, `None`) unless there is an actual reason to do so.
In this case in particular, I see no need to check for it, since it's OK to skip the if block if something like a `False` or a `""` is passed.

Reply:
I'm thinking through this a bit more. I think that either param_type should be a valid `ParamType` or `None`. Otherwise, an error should be raised. I don't see a reason to allow `param_type=False` which makes no sense, and an error should be raised in that case instead of continuing with class instantiation.

I suggest refactoring `ParamType` to be like
```
class ParamType(Enum):
    INT = 1
    # etc.
```

We can use enum on python 3.4+ and use the backported enum34 for python 2.7. Then we can test whether `isinstance(param_type, ParamType)`.

Even further, is there a use case in which `param_type=None` yielding a instance of type `HyperParameter` (rather than a subtype of `HyperParameter`) is acceptable? I can't think what it would be. In this case, perhaps we have
```
if isinstance(param_type, ParamType):
        for subclass in cls.subclasses():
            if subclass.param_type is param_type:
                return super(HyperParameter, cls).__new__(subclass)

raise ValueError('Could not instantiate HyperParameter with param_type {}'.format(param_type))
```

Reply:
Ok, I implemented it as you suggested.
ParamType is now an Enum and passing something other than a valid ParamType, including None, raises a ValueError.

Reply:
Looks great!

For distribution purposes, can we make `enum34` a conditional dependency based on python version?
in both setup.py and requirements.txt, we would have
`'enum34>=1.1.6;python_version==2.7'`